barcode_generator_product_variant/tests/test_barcode_generator_product_variant.py

```diff
@@ -0,0 +1,49 @@
+from odoo.exceptions import UserError
+from odoo.tests.common import SavepointCase
+
+
+class TestProductProductBarcodeWizard(SavepointCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+
+        cls.barcode_rule = cls.env.ref(
+            "barcode_generator_product_variant.product_generated_barcode_demo"
+        )
+        cls.test_product = cls.env.ref("product.product_product_4_product_template")
+
+    def test_generate_missing_barcodes(self):
+        self.assertFalse(
+            any(i for i in self.test_product.mapped("product_variant_ids.barcode"))
```

Comment:
**quibble (non-blocking)**: you don't need the generator here, `mapped` already returns a list

```suggestion
            any(self.test_product.mapped("product_variant_ids.barcode"))
```